### PR TITLE
Resolve some todos in async-await branch

### PR DESCRIPTION
**Pull request by nWacky**
This PR resolves some todos in `async-await` branch.

Changelog:
- Resolve RFC 2565 related todos
- Remove `__juniper_extract_generic` macro
- Add `resolve_into_type_async` to `GraphQLTypeAsync` for asynchronous fragments resolving
- Update `GraphQLTypeAsync` with `async-trait`, update macros
- Resolve "better error message with field/type name" todo, update other panics with query name (query name retrieved by calling `self::name(info)`)
- Rebase onto master 
- Format

### CODE REVIEW

File: examples/warp_async/Cargo.toml [hunk 1]
```diff
@@ -13,7 +13,10 @@ warp = "0.1.19"
 futures-preview = { version = "0.3.0-alpha.19", features = ["async-await", "compat"] }
 reqwest = "0.9.19"
 
-juniper_codegen = { git = "https://github.com/graphql-rust/juniper", branch = "async-await", features = ["async"] }
-juniper = { git = "https://github.com/graphql-rust/juniper", branch = "async-await", features = ["async"] }
-juniper_warp = { git = "https://github.com/graphql-rust/juniper", branch = "async-await", features = ["async"] }
+#juniper_codegen = { git = "https://github.com/graphql-rust/juniper", branch = "async-await", features = ["async"] }
+#juniper = { git = "https://github.com/graphql-rust/juniper", branch = "async-await", features = ["async"] }
+#juniper_warp = { git = "https://github.com/graphql-rust/juniper", branch = "async-await", features = ["async"] }
```
Comment (LegNeato) [inline examples/warp_async/Cargo.toml]: Nit: Remove these

File: juniper/src/macros/tests/args.rs [hunk 2]
```diff
@@ -73,13 +73,12 @@ impl Root {
...
     }
 
-    // TODO: enable once [RFC 2565](https://github.com/rust-lang/rust/issues/60406) is implemented
-    // fn attr_arg_descr(#[doc = "The arg"] arg: i32) -> i32 { 0 }
-    // fn attr_arg_descr_collapse(
-    //     #[doc = "The arg"]
-    //     #[doc = "and more details"]
-    //     arg: i32,
-    // ) -> i32 { 0 }
+    fn attr_arg_descr(arg: i32) -> i32 {
```
Comment (LegNeato) [inline juniper/src/macros/tests/args.rs]: This looks wrong? Also, we have https://github.com/graphql-rust/juniper/pull/441 outstanding

File: juniper/src/macros/tests/union.rs [hunk 3]
```diff
@@ -36,6 +44,16 @@ enum WithGenerics<T> {
...
+enum ResolversFirst {
+    Concrete(Concrete),
+}
+
+enum CommasWithTrailing {
+    Concrete(Concrete),
+}
+enum ResolversWithTrailingComma {
+    Concrete(Concrete),
+}
```
Comment (LegNeato) [inline juniper/src/macros/tests/union.rs]: Why were these added back?

File: juniper/src/macros/tests/args.rs (hunk 2 above)
Comment (nWacky) [reply juniper/src/macros/tests/args.rs]: Got it. 
I put it back to todos and updated a link to #441

File: juniper/src/macros/tests/union.rs (hunk 3 above)
Comment (nWacky) [reply juniper/src/macros/tests/union.rs]: Oops, my mistake

File: juniper/src/macros/tests/field.rs [hunk 4]
```diff
@@ -95,13 +95,16 @@ impl Root {
         Ok(0)
     }
 
+    /*
+     * FIXME: make this work again
```
Comment (LegNeato) [inline juniper/src/macros/tests/field.rs]: This looks like you need to rebase or a bad merge?

File: juniper/src/macros/tests/union.rs (hunk 3 above)
Comment (LegNeato) [reply juniper/src/macros/tests/union.rs]: I don't see this fixed? Is this a bad merge?

**Comment by nWacky** [comment]
Yes, git messed everything up while merging.
Now everything should be good

**Comment by theduke** [comment]
I'm sorry I didn't see this before, but I'm not actually on board with using `async-trait`. 

It's a nice library, but it just adds a bit of syntax sugar. Which isn't really needed when 99% of the code will be auto-generated by the macros.

**Comment by theduke** [comment]
Could you handle reverting, @nWacky ?

**Comment by nWacky** [comment]
@theduke 

Yes, I'll remove async-trait today. Juniper's macros don't depend on async-trait, I wrote all lifetimes by hand

**Comment by theduke** [comment]
Thanks. I think we will want to merge the async branch in the next few days to avoid extra churn.

I also have some bigger refactorings upcoming that could affect you, we might want to chat about that somewhere.

**Comment by nWacky** [comment]
Yeah, that would be great

**Comment by LegNeato** [comment]
@theduke I converted a project at work to async_trait and it made the implementations much more readable as everything didn't have to have `Box` and `Pin` everywhere. It also let me clean up some `futures::future::Either` use as well. But I guess I don't feel strongly enough to keep it.

**Comment by theduke** [comment]
I have a type system change which gets rid of `GraphlQLTypeAsync` that makes this impossible. 

Should be ready today.
